lwt/tls_lwt.ml

```diff
@@ -181,7 +181,11 @@ module Unix = struct
             ( t.state <- inject_state tls t.state ; write_t t tlsdata )
         | None -> Lwt.reraise @@ Invalid_argument "tls: write: socket not ready"
 
-  let write t cs = writev t [cs]
+  let write t ?(off = 0) cs =
+    let buf =
+      if off = 0 then cs else String.sub cs off (String.length cs - off)
+    in
+    writev t [buf]
```

Comment:
This makes a copy which you would not expect from the type :/ but maybe there is room for improvement later? I realize this probably goes  deep into the engine...

Reply:
yes, I disliked this as well. I can see two options: remove the `off`, or improve the performance later. I guess we'll need quite some adjustments to the engine...

Reply:
Maybe it's fine if we add a comment in the documentation that it currently makes a copy. Then, I wonder why not have a `?len` as well? I'm undecided.

Reply:
> Maybe it's fine if we add a comment in the documentation that it currently makes a copy. Then, I wonder why not have a `?len` as well? I'm undecided.

I find adding a documentation is not worth it. It will change. All the other functions as well don't have documentation about potential allocations.

I pushed c001750 which removed the `?off` from `write`. It makes the API very asymmetric (taking writev into account).